midend/lib/Conversion/ConvOptimization/ConvNhwcFhwcOptimizeTile.cpp

```diff
@@ -0,0 +1,342 @@
+    tilingOW = tilingOWParam;
+    tilingOC = tilingOCParam;
+  }
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> /*operands*/,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto convOp = dyn_cast_or_null<mlir::linalg::Conv2DNhwcFhwcOp>(op);
+    auto loc = op->getLoc();
+
+    // Some constant we need.
+    const Value c0 =
+        rewriter.create<arith::ConstantOp>(loc, rewriter.getIndexAttr(0));
+    const Value c1 =
+        rewriter.create<arith::ConstantOp>(loc, rewriter.getIndexAttr(1));
+
+    const Value vecSizeValue =
+        rewriter.create<arith::ConstantOp>(loc, rewriter.getIndexAttr(vecSize));
+    const AffineExpr d0 = rewriter.getAffineDimExpr(0);
+    const AffineExpr d1 = rewriter.getAffineDimExpr(1);
+    const AffineExpr s0 = rewriter.getAffineSymbolExpr(0);
+
+    Value input = op->getOperand(0);
+    Value filter = op->getOperand(1);
+    Value output = op->getOperand(2);
+
+    int strHeight, strWidth, dilHeight, dilWidth;
+
+    // Strides.
+    if (!convOp.getStrides()) {
+      strHeight = 1;
+      strWidth = 1;
+    } else {
+      strHeight = convOp.getStrides().getValues<int64_t>()[0];
+      strWidth = convOp.getStrides().getValues<int64_t>()
+                     [convOp.getStrides().getValues<int64_t>().size() - 1];
+    }
+
+    // Dilations.
+    if (!convOp.getDilations()) {
+      dilHeight = 1;
+      dilWidth = 1;
+    } else {
+      dilHeight = convOp.getDilations().getValues<int64_t>()[0];
+      dilWidth = convOp.getDilations().getValues<int64_t>()
+                     [convOp.getDilations().getValues<int64_t>().size() - 1];
+    }
+
+    ShapedType inputTy = input.getType().cast<ShapedType>();
+    Type elemTy = inputTy.getElementType();
+    VectorType vecTy = VectorType::get(vecSize, elemTy);
+
+    const Value zeroElementType =
+        rewriter.create<arith::ConstantOp>(loc, rewriter.getZeroAttr(elemTy));
+
+    // Dims
+    Value N = rewriter.create<memref::DimOp>(loc, output, 0);  // N
+    Value OH = rewriter.create<memref::DimOp>(loc, output, 1); // OH
+    Value OW = rewriter.create<memref::DimOp>(loc, output, 2); // OW
+    Value OC = rewriter.create<memref::DimOp>(loc, output, 3); // OC
+    Value IC = rewriter.create<memref::DimOp>(loc, input, 3);  // IC
+    Value FH = rewriter.create<memref::DimOp>(loc, filter, 1); // FH
+    Value FW = rewriter.create<memref::DimOp>(loc, filter, 2); // FW
+
+    auto tilingUpperBound =
+        AffineMap::get(2, 1, {d0 + d1, s0}, rewriter.getContext());
+
+    Value stepOH = rewriter.create<affine::AffineApplyOp>(
+        loc, AffineMap::get(1, 0, d0.ceilDiv(tilingOH)), OH);
+    Value stepOW = rewriter.create<affine::AffineApplyOp>(
+        loc, AffineMap::get(1, 0, d0.ceilDiv(tilingOW)), OW);
+    Value stepOC = rewriter.create<affine::AffineApplyOp>(
+        loc, AffineMap::get(1, 0, d0.ceilDiv(tilingOC)), OC);
+
+    // clang format off
+    //  Step 1: Create outer most loops.
```

Comment:
Do we have `Step 2` in this file? If not, the word `Step` had better be removed.